pkg/ring/token_checker.go

```diff
@@ -0,0 +1,253 @@
+package ring
+
+import (
+	"context"
+	"flag"
+	"sync"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+)
+
+// TokenCheckerConfig is the config to configure a TokenChecker.
+type TokenCheckerConfig struct {
+	// Used by ingester when set to true.
+	CheckOnCreate   bool `yaml:"check_on_create,omitempty"`
+	CheckOnAppend   bool `yaml:"check_on_append,omitempty"`
+	CheckOnTransfer bool `yaml:"check_on_transfer,omitempty"`
+
+	// Used by TokenChecker on the interval
+	CheckOnInterval time.Duration `yaml:"check_on_interval"`
+}
+
+// RegisterFlags adds flags required to configure a TokenChecker to
+// the provided FlagSet.
+func (cfg *TokenCheckerConfig) RegisterFlags(f *flag.FlagSet) {
+	cfg.RegisterFlagsWithPrefix("", f)
+}
+
+// RegisterFlagsWithPrefix adds the flags required to config a TokenChecker to
+// the given FlagSet, prefixing each flag with the value provided by prefix.
+func (cfg *TokenCheckerConfig) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.BoolVar(&cfg.CheckOnCreate, prefix+"token-checker.check-on-create", false, "Check that newly created streams fall within expected token ranges")
+	f.BoolVar(&cfg.CheckOnCreate, prefix+"token-checker.check-on-append", false, "Check that existing streams appended to fall within expected token ranges")
+	f.BoolVar(&cfg.CheckOnCreate, prefix+"token-checker.check-on-transfer", false, "Check that streams transferred in using the transfer mechanism fall within expected token ranges")
+	f.DurationVar(&cfg.CheckOnInterval, prefix+"token-checker.check-on-interval", time.Duration(0), "Period with which to check that all in-memory streams fall within expected token ranges. 0 to disable.")
+}
+
+// A TokenChecker is responsible for validating that streams written to an
+// ingester have tokens that fall within an expected range of values.
+// Appropriate values depend on where an ingester's tokens are placed in the
+// ring and what its neighbors are.
+//
+// Checking that newly created streams fall within the expected token
+// ranges ensures that writes to the ingester are distributed properly,
+// while checking for validity of streams on an interval ensures that
+// an ingester's memory only contains appropriate tokens as the ingesters
+// in the ring change over time.
+type TokenChecker struct {
+	// UnexpectedTokenHandler is invoked by the TokenChecker whenever
+	// CheckAllTokens is called, even when no unexpected tokens are
+	// found. If nil, is a no-op.
+	UnexpectedTokenHandler func(tokens []uint32)
+
+	mut        sync.Mutex
```

Comment:
What I actually meant was to group it together with `ring` and `expectedRanges`, because it protects those two fields.

```
	mut            sync.Mutex
	ring           *Desc
	expectedRanges []TokenRange
```

Reply:
This code is still little confused about mutex and whether to use it or not for context and cancel function. Shutdown uses mutex for calling cancel(), but Start doesn't use mutex for setting it.

(I liked previous design better, when NewTokenChecker created context and started go-routines as well... there was no need to split that, except for setting UnexpectedTokenHandler, but that could have been passed to NewTokenChecker method instead. Then mutex wouldn't be necessary for context/cancel).